### ISSUE TITLE
Query Predicate: Incorrect comparison logic with real numbers.

### ISSUE THREAD

**Issue by ScottNicol** [hazelcast/hazelcast#6188]
This issue is present in 3.5.2.

Predicates `BetweenPredicate`, `GreaterLessPredicate` and `EqualPredicate` use `compareTo()` and `equals()` to compare two values. For `float` and `double` this results in unexpected behavior. See java documentation for [`compareTo`](http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html#compareTo%28java.lang.Double%29) and [`equals`](http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html#equals%28java.lang.Object%29).

There are two problems.

-0.0 and 0.0 are equal when using java operators, but not equal with compareTo() and equals().  If you have a query x < 0, then -0.0 will match.

NaN is not equal to anything (including itself). So NaN == NaN should be false and NaN != NaN should be true. However with compareTo() and equals() NaN is assigned for the purposes of ordering as greater than infinity. This means that NaN = NaN is true, NaN != NaN is false, and x > 0, or even x > infinity, is true if x is NaN.

The fix is to check if the entryValue class is `Double` or `Float`, and if so cast the entryValue and value to the appropriate type, use `doubleValue()` (or `floatValue()` if float) to get the primitive values, then compare using java operators.

**Comment by ScottNicol** [comment]
Proposed patch:

```
diff --git a/hazelcast-3.5.2/hazelcast/src/main/java/com/hazelcast/query/Predicates.java b/hazelcast-3.5.2/hazelcast/src/main/java/com/hazelcast/query/Predicates.java
index 787c74b..584cc74 100644
--- a/hazelcast-3.5.2/hazelcast/src/main/java/com/hazelcast/query/Predicates.java
+++ b/hazelcast-3.5.2/hazelcast/src/main/java/com/hazelcast/query/Predicates.java
@@ -152,7 +152,19 @@ public final class Predicates {
             if (fromConvertedValue == null || toConvertedValue == null) {
                 return false;
             }
-            return entryValue.compareTo(fromConvertedValue) >= 0 && entryValue.compareTo(toConvertedValue) <= 0;
+            if (entryValue.getClass() == Double.class) {
+              double entryValueD = ((Double)entryValue).doubleValue();
+              double fromConvertedValueD = ((Double)fromConvertedValue).doubleValue();
+              double toConvertedValueD = ((Double)toConvertedValue).doubleValue();
+              return (entryValueD >= fromConvertedValueD) && (entryValueD <= toConvertedValueD);
+            } else if (entryValue.getClass() == Float.class) {
+              float entryValueF = ((Float)entryValue).floatValue();
+              float fromConvertedValueF = ((Float)fromConvertedValue).floatValue();
+              float toConvertedValueF = ((Float)toConvertedValue).floatValue();
+              return (entryValueF >= fromConvertedValueF) && (entryValueF <= toConvertedValueF);
+            } else {
+              return entryValue.compareTo(fromConvertedValue) >= 0 && entryValue.compareTo(toConvertedValue) <= 0;
+            }
         }
 
         @Override
@@ -675,7 +687,18 @@ public final class Predicates {
                 return false;
             }
             final Comparable attributeValue = convert(mapEntry, entryValue, value);
-            final int result = entryValue.compareTo(attributeValue);
+            final int result;
+            if (entryValue.getClass() == Double.class) {
+              double entryValueD = ((Double)entryValue).doubleValue();
+              double attributeValueD = ((Double)attributeValue).doubleValue();
+              result = (entryValueD == attributeValueD) ? 0 : (entryValueD < attributeValueD) ? -1 : 1;
+            } else if (entryValue.getClass() == Float.class) {
+              float entryValueF = ((Float)entryValue).floatValue();
+              float attributeValueF = ((Float)attributeValue).floatValue();
+              result = (entryValueF == attributeValueF) ? 0 : (entryValueF < attributeValueF) ? -1 : 1;
+            } else {
+              result = entryValue.compareTo(attributeValue);
+            }
             return equal && result == 0 || (less ? (result < 0) : (result > 0));
         }
 
@@ -786,7 +809,13 @@ public final class Predicates {
                 return value == null || value == IndexImpl.NULL;
             }
             value = convert(mapEntry, entryValue, value);
-            return entryValue.equals(value);
+            if (entryValue.getClass() == Double.class) {
+              return ((Double)entryValue).doubleValue() == ((Double)value).doubleValue();
+            } else if (entryValue.getClass() == Float.class) {
+              return ((Float)entryValue).floatValue() == ((Float)value).floatValue();
+            } else {
+              return entryValue.equals(value);
+            }
         }
 
         @Override
```

**Comment by jerrinot** [comment]
@ScottNicol: this is an interesting realization! The suggested patch would solve it only partially though. When an index is used then it does not call the `apply()` method. When searching in index it's obviously using the `compareTo()`/`equals()` methods. 

I'm actually not sure whether we should attempt to fix this or just document it properly. As fixing this would break the current rule that `equals()`/`compareTo()` are used for comparison.

**Comment by ScottNicol** [comment]
compareTo()/equals() is written this way to provide an ordering for things like hashtable keys, but it's not a numeric comparison. I think it's fine to keep indexing using compareTo()/equals().

It was interested to figure this bug out. I got a question about why so many 0's were making it past a query "x < 0". And then there's NaN, which showed up in queries like "x > 100".

The behavior of -0.0 vs 0.0 and NaN vs numbers are spec'ed in IEEE 754, which is the floating point standard that is used on almost every system and programming language. So hazelcast produces mathematically different results than Java, C++, C#, etc.

```
class Compare {
  public static void main(String[] args) {
    double negZero = 0.0D * -1;
    double zero = 0.0D * 1;
    double nan = Double.NaN;

    System.out.println("compare(negZero,zero): " + Double.compare(negZero, zero));
    System.out.println("negZero < zero: " + (negZero < zero));
    System.out.println("negZero == zero: " + (negZero == zero));
    System.out.println("compare(nan, nan): " + Double.compare(nan, nan));
    System.out.println("nan == nan: " + (nan == nan));
    System.out.println("nan != nan: " + (nan != nan));
    System.out.println("compare(nan, 0): " + Double.compare(nan, 0));
    System.out.println("nan > 0: " + (nan > 0));
  }
}
```

results in:

```
compare(negZero,zero): -1
negZero < zero: false
negZero == zero: true
compare(nan, nan): 0
nan == nan: false
nan != nan: true
compare(nan, 0): 1
nan > 0: false
```

**Comment by jerrinot** [comment]
@ScottNicol: if we change the behaviour when index is not used (=the `apply()` method) then obviously we'd need to adjust behaviour when index is used too. Otherwise creating an index would lead to different results and obviously that's not desired. 

From the other hand I understand your reasoning and I do agree the current situation is not ideal. Let's get more people involved. @ajermakovics, @gurbuzali: WDYT?

**Comment by ScottNicol** [comment]
I have an idea. How about doing both? <fp-compare-with-operators enabled=false> or something like that. Make the default the current behavior, users can switch on if needed. If at some point in the future it's decided that operators are preferred then the default could be flipped. Whatever the default is, users can choose what's right for them.

**Comment by jerrinot** [comment]
@ScottNicol: A switch is a possibility. However each switch makes Hazelcast harder to use for both development and operations. Last but not least: It increases our maintenance burden and likely-hood of a bug (it's hard to test test every possible permutation)

**Comment by pveentjer** [comment]
I have removed QuSP label since this is a logical error and related to map. So it is a core issue.

**Comment by mdogan** [comment]
@jerrinot: is this planned for 3.6 or should we postpone to 3.7?

**Comment by jerrinot** [comment]
let's postpone it into 3.7

**Comment by tombujok** [comment]
Wasn't that fixed by @cpiotr ?

**Comment by tombujok** [comment]
@ScottNicol Thx for discovering this. 
I've documented how it fails here #9631 
In order not to break the backward compatibility we will fix it in 4.0

**Comment by mmedenjak** [comment]
We introduced many improvements in this are with 3.12:
https://github.com/hazelcast/hazelcast/pull/14597

There are still some cases where it may not work correctly but maybe we can close this issue for now and reopen a new one as new cases occur.
